gateway/src/metrics.rs

```diff
@@ -0,0 +1,1069 @@
+                    slog::trace!(
+                        &self.log,
+                        "polled SP, no samples returned";
+                        "num_samples" => 0usize
+                    );
+                }
+                Ok(samples) => {
+                    slog::trace!(
+                        &self.log,
+                        "polled SP successfully";
+                        "num_samples" => samples.len(),
+                    );
+
+                    if let Err(_) = self.sample_tx.send(samples) {
+                        slog::debug!(
+                            &self.log,
+                            "all sample receiver handles have been dropped! \
+                             presumably we are shutting down...";
+                        );
+                        return Ok(());
+                    }
+                }
+                // No SP is currently present for this ID. This may change in
+                // the future: a cubby that is not populated at present may have
+                // a sled added to it in the future. So, let's wait until it
+                // changes.
+                Err(CommunicationError::NoSpDiscovered) => {
+                    slog::info!(
+                        &self.log,
+                        "no SP is present for this slot. waiting for a \
+                         little buddy to appear...";
+                    );
+                    let mut watch = sp.sp_addr_watch().clone();
+                    loop {
+                        if let Some((addr, port)) = *watch.borrow_and_update() {
+                            // Ladies and gentlemen...we got him!
+                            slog::info!(
+                                &self.log,
+                                "found a SP, resuming polling.";
+                                "sp_addr" => ?addr,
+                                "sp_port" => ?port,
+                            );
+                            break;
+                        }
+
+                        // Wait for an address to be discovered.
+                        slog::debug!(&self.log, "waiting for a SP to appear.");
+                        if watch.changed().await.is_err() {
+                            slog::debug!(
+                                &self.log,
+                                "SP address watch has been closed, presumably \
+                                 we are shutting down";
+                            );
+                            return Ok(());
+                        }
+                    }
+                }
+                Err(error) => {
+                    slog::warn!(
+                        &self.log,
+                        "failed to poll SP, will try again momentarily...";
+                        "error" => %error,
+                    );
+                    // TODO(eliza): we should probably have a metric for failed
+                    // SP polls.
```

Comment:
I was going to ask if that's what `ComponentMetrics::poll_error()` was for, but that's for polling errors on a _specific_ component, right?

How urgently do you want to tackle this TODO? (If not soon, probably worth filing an issue for someone to come back and fill it in?)

Reply:
Yeah, the `poll_error_count` metric has the `hardware_component` target, so it's specific to the individual component. A metric counting the total number of  poll errors for the whole SP would require an additional metric schema, so I was hoping to add it in a subsequent branch.

Reply:
Works for me :+1: